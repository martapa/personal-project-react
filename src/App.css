#root > div {
  margin: 0 auto;
  max-width: 1280px;
  padding: 0 15px 15px;
}

#root label {
  display: block;
  margin-top: 15px;
}

#root input[type="text"] {
  border: 5px solid #000;
  display: block;
  margin-top: 15px;
  padding: 15px;
}

#root h2 {
  margin: 15px 0 0;
}

#root .row div {
  border: 5px solid #000;
  margin-top: 15px;
  padding: 30px;
}

#root h3 {
  margin: 0;
}

#root p {
  margin-bottom: 0;
}

#root button,
#root input[type="submit"] {
  background: #000;
  border: 0;
  color: #fff;
  font-size: 1.5rem;
  font-weight: bold;
  letter-spacing: 0.05rem;
  margin-top: 15px;
  padding: 15px;
  text-transform: uppercase;
}