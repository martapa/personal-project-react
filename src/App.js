import React, { useState } from "react";
import axios from "axios";
import "./App.css";


const Forked = ({ arr }) => {
  const forkedList = arr.map((fork, index) => (
    <div key={`${fork.full_name}-${index}`}>
      <a href={fork.clone_url}>
        <h3>{fork.full_name}</h3>
        <p>{`Forked from: ${fork.parent_name}`}</p>
      </a>
    </div>
  ));

  return <div className="row">{forkedList}</div>;
};

const PullEvents = ({ arr }) => {
  const pullList = arr.map(pull => (
    <div key={pull.repo_name}>
      <a href={pull.url}>
        <h3>{pull.repo_name}</h3>
        <p>{`Status: ${pull.status}`}</p>
      </a>
    </div>
  ));

  return <div className="row">{pullList}</div>;
};

function App() {
  const [githubUser, setUser] = useState("pkanal");
  const [repos, setRepos] = useState([]);
  const [events, setEvents] = useState([]);

  const handleBackButton = e => {
    setRepos([]);
    setEvents([]);
  };

  const handleUserChange = e => {
    setUser(e.target.value);
  };

  const handleUserSubmit = async e => {
    e.preventDefault();

    try {
      const resRepos = await axios.get(
        `https://api.github.com/users/${githubUser}/repos?sort=created`
      );
      const recentRepos = resRepos.data.slice(0, 2);
      const parents = [];
      for (let i = 0; i < recentRepos.length; i++) {
        const resParent = await axios.get(
          `https://api.github.com/repos/${githubUser}/${recentRepos[i].name}`
        );
        const parentName = resParent.data.parent
          ? resParent.data.parent.name
          : "No data avalible";
        const newRepo = {
          username: githubUser,
          full_name: recentRepos[i].full_name,
          clone_url: recentRepos[i].clone_url,
          parent_name: parentName
        };
        parents.push(newRepo);
      }
      setRepos(parents);

      const resEvents = await axios.get(
        `https://api.github.com/users/${githubUser}/events`
      );

      console.log(resEvents);
      const pullEvents = resEvents.data.filter(
        event => event.type === "PullRequestEvent"
      );

      const newPullEvents = pullEvents.map(event => {
        const newPullEvent = {
          repo_name: event.repo.name,
          status: event.payload.action,
          url: event.repo.url
        };
        return newPullEvent;
      });

      setEvents(newPullEvents);
    } catch (err) {
      console.log(err);
    }
  };

  return (
    <>
      {repos.length === 0 ? (
        <form onSubmit={handleUserSubmit}>
          <label>
            Github username:
            <input
              type="text"
              name="name"
              value={githubUser}
              onChange={handleUserChange}
            />
          </label>
          <input type="submit" value="Submit" />
        </form>
      ) : (
        <div>
          <h2>Recent forks</h2>
          <Forked arr={repos} />
          <h2>Recent pull requests</h2>
          <PullEvents arr={events} />
          <button onClick={handleBackButton}>Back button</button>
        </div>
      )}
    </>
  );
}

export default App;
