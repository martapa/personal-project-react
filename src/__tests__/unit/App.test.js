import React from "react";
import ReactDOM from "react-dom";
import App from "../../App";
import axios from "axios";
import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import toJson from "enzyme-to-json";

Enzyme.configure({ adapter: new Adapter() });

import { shallow } from "enzyme";

import repos from "../mockData/repos.json";

axios.get = jest.fn(() => Promise.resolve(repos));

it("calls axios and returns repos", async () => {
  const response = await axios.get(
    "https://api.github.com/users/martapa/repos?sort=created"
  );
  expect(response).toEqual(repos);
  expect(axios.get).toHaveBeenCalledTimes(1);
  expect(axios.get).toHaveBeenCalledWith(
    "https://api.github.com/users/martapa/repos?sort=created"
  );
});

describe("<App /> shallow rendering", () => {
  it("should render App", () => {
    const wrapper = shallow(<App />);
  });

  it("matches the snapshot", () => {
    const wrapper = shallow(<App />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
