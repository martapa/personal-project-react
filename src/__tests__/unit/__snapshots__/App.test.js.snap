// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> shallow rendering matches the snapshot 1`] = `
<Fragment>
  <form
    onSubmit={[Function]}
  >
    <label>
      Github username:
      <input
        name="name"
        onChange={[Function]}
        type="text"
        value="pkanal"
      />
    </label>
    <input
      type="submit"
      value="Submit"
    />
  </form>
</Fragment>
`;
